### PR TITLE
[Telink] Add Lock, Pump & Pump Controller demo-apps

### PR DESCRIPTION
**Pull request by s07641069**
**Problem**

These changes add new Telink app examples

**Change overview**

- Added Telink Lock example
- Added Telink Pump example
- Added Telink Pump Controller example

**Testing**

Tested manually with chip-tool.

Steps:
- Run: $ chip-tool pairing ble-thread <...>
- Wait till success

### CODE REVIEW

File: examples/lock-app/telink/src/BoltLockManager.cpp [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    return true;
+}
+
+bool BoltLockManager::GetCredential(uint16_t credentialIndex, CredentialTypeEnum credentialType,
+                                    EmberAfPluginDoorLockCredentialInfo & credential) const
+{
+    VerifyOrReturnError(credentialIndex > 0 && credentialIndex <= CONFIG_LOCK_NUM_CREDENTIALS, false);
+
+    credential = mCredentials[credentialIndex - 1];
```
Comment (bzbarsky-apple) [inline examples/lock-app/telink/src/BoltLockManager.cpp:84]: This does not implement the spec correctly.  Credential indices are per-type, not global.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/lock-app/telink/src/BoltLockManager.cpp (hunk 1 above)
Comment (interfer) [reply examples/lock-app/telink/src/BoltLockManager.cpp:84]: @bzbarsky-apple If I got it right, do we need to operate only with PIN in the lock-app as it's done in #24532?

Comment (bzbarsky-apple) [reply examples/lock-app/telink/src/BoltLockManager.cpp:84]: I'm not following.  #24532 made the Linux lock-app implement the spec properly.

What you should do is follow the spec.  Your FeatureMap says you support PIN and RFID, so you need to correctly implement supporting both.

If you want to only support PIN, you can do that, but then you need to change your FeatureMap accordingly.

Comment (interfer) [reply examples/lock-app/telink/src/BoltLockManager.cpp]: Thanks, updated with following commit. AFAIK `0x101` in FeatureMap will use `kUser` and `kPinCredential`, accornigly to "Application Clusters 5.2.2. Features", do You agree @bzbarsky-apple ?

File: examples/lock-app/telink/include/BoltLockManager.h [hunk 2]
```diff
@@ -99,8 +106,16 @@ class BoltLockManager
...
+    struct LockCredentialInfo
+    {
+        DlCredentialStatus status;
+        CredentialTypeEnum credentialType;
+        chip::FabricIndex createdBy;
+        chip::FabricIndex modifiedBy;
+        uint8_t credentialData[CONFIG_LOCK_CREDENTIAL_INFO_MAX_TYPES];
+        size_t credentialDataSize;
+    };
+    std::vector<std::vector<LockCredentialInfo>> mCredentials;
```
Comment (bzbarsky-apple) [inline examples/lock-app/telink/include/BoltLockManager.h:118]: So what gives this a nonzero capacity?  There are various capacity tests in the added code, as far as I can tell....

Is this app able to pass https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/door_lock_Cluster.adoc#tc-drlk-2-11-verification-for-multiple-credential-types-dut-server for whatever feature map it has?

Comment (interfer) [reply examples/lock-app/telink/include/BoltLockManager.h:121]: Thanks, `mCredentials` is allocated in constructor for now. Actually we have not run any certification tests yet as we don't have plans for certification in nearest time. So these apps have quite a short functionality(demo) for the present

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/pump-controller-app/telink/src/ZclCallbacks.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/ConcreteAttributePath.h>
+
+using namespace ::chip;
+using namespace ::chip::app::Clusters;
+
+void MatterPostAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t type, uint16_t size,
+                                       uint8_t * value)
+{
+    if (attributePath.mClusterId == OnOff::Id && attributePath.mAttributeId == OnOff::Attributes::OnOff::Id)
```
Comment (bzbarsky-apple) [inline examples/pump-controller-app/telink/src/ZclCallbacks.cpp:32]: I don't understand this part. This app does not implement an On/Off server, so how would this code ever be triggered?

Comment (interfer) [reply examples/pump-controller-app/telink/src/ZclCallbacks.cpp]: Removed unused code

File: examples/pump-controller-app/telink/src/ZclCallbacks.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ * @param endpoint   Ver.: always
+ *
+ * TODO Issue #3841
+ * emberAfOnOffClusterInitCallback happens before the stack initialize the cluster
+ * attributes to the default value.
+ * The logic here expects something similar to the deprecated Plugins callback
+ * emberAfPluginOnOffClusterServerPostInitCallback.
+ *
+ */
+void emberAfOnOffClusterInitCallback(EndpointId endpoint)
```
[the remark this answers is not in the document]
Comment (interfer) [reply examples/pump-controller-app/telink/src/ZclCallbacks.cpp]: As above, cut out not used callbacks

**Review by andy31415 — APPROVED** [review]

**Pull request merged**